File: FBDeviceControl/Management/FBAMRestorableDeviceManager.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBAMRestorableDeviceManager.h"

#import "FBDeviceControlError.h"
#import "FBDeviceControlFrameworkLoader.h"
#import "FBAMRestorableDevice.h"

@interface FBAMRestorableDeviceManager ()

@property (nonatomic, strong, readonly) dispatch_queue_t queue;
@property (nonatomic, assign, readonly) AMDCalls calls;
@property (nonatomic, assign, readwrite) int registrationID;
@property (nonatomic, copy, readonly) NSString *ecidFilter;

- (NSDictionary<NSString *, id> *)infoForRestorableDevice:(AMRestorableDeviceRef)device;

@end

static NSString *NotificationTypeToString(AMRestorableDeviceNotificationType status)
{
  switch (status) {
    case AMRestorableDeviceNotificationTypeConnected:
      return @"connected";
    case AMRestorableDeviceNotificationTypeDisconnected:
      return @"disconnected";
    default:
      return @"unknown";
  }
}

static void FB_AMRestorableDeviceListenerCallback(AMRestorableDeviceRef device, AMRestorableDeviceNotificationType status, void *context)
{
  FBAMRestorableDeviceManager *manager = (__bridge FBAMRestorableDeviceManager *)context;
  id<FBControlCoreLogger> logger = manager.logger;
  AMRestorableDeviceState deviceState = manager.calls.RestorableDeviceGetState(device);
  FBiOSTargetState targetState = [FBAMRestorableDevice targetStateForDeviceState:deviceState];
  NSString *identifier = [@(manager.calls.RestorableDeviceGetECID(device)) stringValue];
  [logger logFormat:@"%@ %@ in state %@", device, NotificationTypeToString(status), FBiOSTargetStateStringFromState(targetState)];
  if (manager.ecidFilter && ![identifier isEqualToString:manager.ecidFilter]) {
    [logger logFormat:@"Ignoring %@ as it does not match filter of %@", device, manager.ecidFilter];
    return;
  }
  switch (status) {
    case AMRestorableDeviceNotificationTypeConnected: {
      NSDictionary<NSString *, id> *info = [manager infoForRestorableDevice:device];
      [logger logFormat:@"Caching restorable device values %@", info];
      [manager deviceConnected:device identifier:identifier info:info];
      return;
    }
    case AMRestorableDeviceNotificationTypeDisconnected:
      [manager deviceDisconnected:device identifier:identifier];
      return;
    default:
      [logger logFormat:@"Unknown Restorable Notification %d", status];
      return;
  }
}

@implementation FBAMRestorableDeviceManager

#pragma mark Initializers

- (instancetype)initWithCalls:(AMDCalls)calls queue:(dispatch_queue_t)queue ecidFilter:(NSString *)ecidFilter logger:(id<FBControlCoreLogger>)logger
{
  self = [super initWithLogger:logger];
  if (!self) {
    return nil;
  }

  _calls = calls;
  _queue = queue;
  _ecidFilter = ecidFilter;

  return self;
}

#pragma mark Abstract Implementation

- (BOOL)startListeningWithError:(NSError **)error
{
  int registrationID = self.calls.RestorableDeviceRegisterForNotifications(
    FB_AMRestorableDeviceListenerCallback,
    (void *) CFBridgingRetain(self),
    0,
    0
  );
  if (registrationID < 1) {
    return [[FBDeviceControlError
      describeFormat:@"AMRestorableDeviceRegisterForNotifications failed with %d", registrationID]
      failBool:error];
  }
  self.registrationID = registrationID;
  return YES;
}

- (BOOL)stopListeningWithError:(NSError **)error
{
  int registrationID = self.registrationID;
  self.registrationID = 0;
  if (registrationID < 1) {
    return [[FBDeviceControlError
      describe:@"Cannot unregister from AMRestorableDevice notifications, no subscription"]
      failBool:error];
  }

  // Return of AMRestorableDeviceUnregisterForNotifications seems to be some random number.
  // However, giving invalid registrationID is fine and we still get logging.
  self.calls.RestorableDeviceUnregisterForNotifications(registrationID);
  return YES;
}

- (FBAMRestorableDevice *)constructPublic:(AMRestorableDeviceRef)privateDevice identifier:(NSString *)identifier info:(NSDictionary<NSString *,id> *)info
{
  return [[FBAMRestorableDevice alloc] initWithCalls:self.calls restorableDevice:privateDevice allValues:info logger:[self.logger withName:identifier]];
}

+ (void)updatePublicReference:(FBAMRestorableDevice *)publicDevice privateDevice:(AMRestorableDeviceRef)privateDevice identifier:(NSString *)identifier info:(NSDictionary<NSString *,id> *)info
{
  publicDevice.restorableDevice = privateDevice;
  publicDevice.allValues = info;
}

+ (AMRestorableDeviceRef)extractPrivateReference:(FBAMRestorableDevice *)publicDevice
{
  return publicDevice.restorableDevice;
}

- (NSDictionary<NSString *, id> *)infoForRestorableDevice:(AMRestorableDeviceRef)device
{
  return @{
    FBDeviceKeyChipID: @(self.calls.RestorableDeviceGetChipID(device)),
    FBDeviceKeyDeviceClass: @(self.calls.RestorableDeviceGetDeviceClass(device)),
    FBDeviceKeyLocationID: @(self.calls.RestorableDeviceGetLocationID(device)),
    FBDeviceKeySerialNumber: CFBridgingRelease(self.calls.RestorableDeviceCopySerialNumber(device)) ?: NSNull.null,
    FBDeviceKeyDeviceName: CFBridgingRelease(self.calls.RestorableDeviceCopyUserFriendlyName(device)) ?: NSNull.null,
    FBDeviceKeyProductType: CFBridgingRelease(self.calls.RestorableDeviceCopyProductString(device)) ?: NSNull.null,
    FBDeviceKeyUniqueChipID: @(self.calls.RestorableDeviceGetECID(device)),
  };
}

@end

File: FBDeviceControl/Management/FBAMRestorableDeviceManager.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

private func notificationTypeDescription(_ status: AMRestorableDeviceNotificationType) -> String {
  switch status {
  case .connected:
    return "connected"
  case .disconnected:
    return "disconnected"
  @unknown default:
    return "unknown"
  }
}

/// The C callback MobileDevice delivers restorable device notifications to.
///
/// A C function cannot capture, so the manager travels as the context pointer that
/// `startListening` registers. It is bridged back unretained: the registration holds the only
/// retain, and gives it back when it is torn down.
private func restorableDeviceListenerCallback(
  _ device: AMRestorableDevice?,
  _ status: AMRestorableDeviceNotificationType,
  _ context: UnsafeMutableRawPointer?
) {
  guard let device, let context else {
    return
  }
  let manager = Unmanaged<FBAMRestorableDeviceManager>.fromOpaque(context).takeUnretainedValue()
  manager.handleNotification(device: device, status: status)
}

/// Obtains `FBAMRestorableDevice` instances.
@objc(FBAMRestorableDeviceManager)
public final class FBAMRestorableDeviceManager: FBDeviceManager<FBAMRestorableDevice> {

  // MARK: - Properties

  private let calls: AMDCalls
  private let workQueue: DispatchQueue
  private let asyncQueue: DispatchQueue
  private let ecidFilter: String?
  private var registrationID: Int32 = 0
  private var notificationContext: UnsafeMutableRawPointer?

  // MARK: - Initializers

  public init(
    calls: AMDCalls,
    work workQueue: DispatchQueue,
    asyncQueue: DispatchQueue,
    ecidFilter: String?,
    logger: any FBControlCoreLogger
  ) {
    self.calls = calls
    self.workQueue = workQueue
    self.asyncQueue = asyncQueue
    self.ecidFilter = ecidFilter
    super.init(logger: logger)
  }

  // MARK: - Notifications

  fileprivate func handleNotification(device: AMRestorableDevice, status: AMRestorableDeviceNotificationType) {
    // Unrecognised values fall to `.unknown`, which `targetState(for:)` maps the same way its
    // `default` did. `.DFU` would report a real state the device is not in.
    let deviceState = AMRestorableDeviceState(rawValue: calls.RestorableDeviceGetState(device)) ?? .unknown
    let targetState = FBAMRestorableDevice.targetState(for: deviceState)
    let identifier = String(calls.RestorableDeviceGetECID(device))
    logger.log(
      "\(device) \(notificationTypeDescription(status)) in state \(FBiOSTargetStateStringFromState(targetState).rawValue)")

    if let ecidFilter, identifier != ecidFilter {
      logger.log("Ignoring \(device) as it does not match filter of \(ecidFilter)")
      return
    }

    switch status {
    case .connected:
      let info = Dictionary(uniqueKeysWithValues: info(forRestorableDevice: device).map { ($0.key.rawValue, $0.value) })
      logger.log("Caching restorable device values \(info)")
      deviceConnected(device, identifier: identifier, info: info)
    case .disconnected:
      deviceDisconnected(device, identifier: identifier)
    @unknown default:
      logger.log("Unknown Restorable Notification \(status.rawValue)")
    }
  }

  // MARK: - Abstract Implementation

  public override func startListening() throws {
    // Retained for as long as the registration lasts: the callback bridges this back unretained, so
    // the manager must not be deallocated while MobileDevice can still deliver a notification.
    // Balanced in `stopListening`, and below if the registration itself fails.
    let context = Unmanaged.passRetained(self).toOpaque()
    let registrationID = calls.RestorableDeviceRegisterForNotifications(
      restorableDeviceListenerCallback,
      context,
      0,
      0)
    guard registrationID >= 1 else {
      Unmanaged<FBAMRestorableDeviceManager>.fromOpaque(context).release()
      throw FBAMRestorableDeviceManagerError.registrationFailed(status: registrationID)
    }
    self.registrationID = registrationID
    self.notificationContext = context
  }

  public override func stopListening() throws {
    let registrationID = self.registrationID
    self.registrationID = 0
    guard registrationID >= 1 else {
      throw FBAMRestorableDeviceManagerError.notRegistered
    }

    // The return of AMRestorableDeviceUnregisterForNotifications seems to be some random number.
    // However, giving an invalid registrationID is fine and we still get logging.
    _ = calls.RestorableDeviceUnregisterForNotifications(registrationID)
    // No further notification can be delivered, so the callback's context can be given back.
    if let context = notificationContext {
      notificationContext = nil
      Unmanaged<FBAMRestorableDeviceManager>.fromOpaque(context).release()
    }
  }

  public override func constructPublic(
    _ privateDevice: CFTypeRef,
    identifier: String,
    info: [String: Any]?
  ) -> FBAMRestorableDevice {
    FBAMRestorableDevice(
      calls: calls,
      restorableDevice: privateDevice,
      allValues: info ?? [:],
      work: workQueue,
      asyncQueue: asyncQueue,
      logger: logger.withName(identifier))
  }

  public override class func updatePublicReference(
    _ publicDevice: FBAMRestorableDevice,
    privateDevice: CFTypeRef,
    identifier: String,
    info: [String: Any]?
  ) {
    publicDevice.restorableDevice = privateDevice
    publicDevice.allValues = info ?? [:]
  }

  public override class func extractPrivateReference(_ publicDevice: FBAMRestorableDevice) -> Unmanaged<AnyObject>? {
    Unmanaged.passUnretained(publicDevice.restorableDevice)
  }

  // MARK: - Private

  private func info(forRestorableDevice device: AMRestorableDevice) -> [FBDeviceKey: Any] {
    [
      FBDeviceKey.chipID: Int(calls.RestorableDeviceGetChipID(device)),
      FBDeviceKey.deviceClass: Int(calls.RestorableDeviceGetDeviceClass(device)),
      FBDeviceKey.locationID: Int(calls.RestorableDeviceGetLocationID(device)),
      FBDeviceKey.serialNumber: calls.RestorableDeviceCopySerialNumber(device)?.takeRetainedValue() as String? ?? NSNull(),
      FBDeviceKey.deviceName: calls.RestorableDeviceCopyUserFriendlyName(device)?.takeRetainedValue() as String? ?? NSNull(),
      FBDeviceKey.productType: calls.RestorableDeviceCopyProductString(device)?.takeRetainedValue() as String? ?? NSNull(),
      FBDeviceKey.uniqueChipID: NSNumber(value: calls.RestorableDeviceGetECID(device)),
    ]
  }
}

public enum FBAMRestorableDeviceManagerError: Error, LocalizedError {
  case registrationFailed(status: Int32)
  case notRegistered

  public var errorDescription: String? {
    switch self {
    case let .registrationFailed(status):
      return "AMRestorableDeviceRegisterForNotifications failed with \(status)"
    case .notRegistered:
      return "Cannot unregister from AMRestorableDevice notifications, no subscription"
    }
  }
}
